src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -1546,6 +1551,143 @@ namespace stan {
     boost::phoenix::function<validate_integrate_ode_control>
     validate_integrate_ode_control_f;
 
+    template <class T>
+    void validate_algebra_solver_non_control_args(const T& alg_fun,
+                                                  const variable_map& var_map,
+                                                  bool& pass,
+                                                  std::ostream& error_msgs) {
+      pass = true;
+      // test function argument type
+      expr_type sys_result_type(VECTOR_T, 0);
+      std::vector<expr_type> sys_arg_types;
+      sys_arg_types.push_back(expr_type(VECTOR_T, 0));  // x
+      sys_arg_types.push_back(expr_type(VECTOR_T, 0));  // y
+      sys_arg_types.push_back(expr_type(DOUBLE_T, 1));  // dat
+      sys_arg_types.push_back(expr_type(INT_T, 1));  // dat_int
+      function_signature_t system_signature(sys_result_type, sys_arg_types);
+      if (!function_signatures::instance()
+          .is_defined(alg_fun.system_function_name_, system_signature)) {
+        error_msgs << "first argument to "
+                   << "algebra_solver"
+                   << " must be the name of a function with signature"
+                   << " (vector, vector, real[], int[]) : vector ";
+        pass = false;
```

Comment:
I think it's helpful to see all the errors in one go. And these errors still correspond to "one spot", since they are all caused by one call to the function.